File: python/test/test_symmetric.py
# SPDX-License-Identifier: Apache-2.0
#
# Copyright (C) 2017-2026 
# Dave Deriso <dderiso@alumni.stanford.edu>, Twitter: @davederiso
# Stephen Boyd
#
# GDTW is a Python/C++ library that performs dynamic time warping.
# GDTW improves upon other methods (such as the original DTW, ShapeDTW, and FastDTW) by introducing regularization,
# which obviates the need for pre-processing, and cross-validation for choosing optimal regularization hyper-parameters.
#
# Paper: https://rdcu.be/cT5dD
# Source: https://github.com/dderiso/gdtw
# Docs: https://dderiso.github.io/gdtw

"""
Symmetric warping path: ψ(t) = 2t − φ(t) computed inside the same DP solve.

Asymmetric mode reads y at t_i and varies x at Tau[i, j]. Symmetric mode keeps
the same DP, but reads y at ψ(t_i, j) = 2 t_i − Tau[i, j] so the residual
couples both directions.

End-to-end shape and identity-warp behavior is covered by parametrized cases
in test_warp.py; this file pins the symmetric-only invariants (ψ accessor,
ψ + φ = 2t, default-is-asymmetric).
"""
import numpy as np
import pytest

import gdtw

from _helpers import scalar_pair


def test_symmetric_exposes_psi_accessors():
    _, x, y = scalar_pair(T=200)
    _, _, _, g = gdtw.warp(x, y, params={"symmetric": True})
    grid = np.linspace(0, 1, 50)
    np.testing.assert_allclose(g.psi(grid), 2.0 * grid - g.phi(grid), atol=1e-12)
    assert g.get_psi_values().shape == (g.N,)


def test_symmetric_psi_complementary():
    """ψ(t) + φ(t) == 2t at every node, by construction."""
    _, x, y = scalar_pair(T=200)
    _, _, _, g = gdtw.warp(x, y, params={"symmetric": True})
    np.testing.assert_allclose(g.get_psi_values() + g.tau, 2.0 * g.t, atol=1e-12)


def test_symmetric_default_is_asymmetric():
    """Omitting `symmetric` must reproduce the legacy asymmetric path bit-identically."""
    _, x, y = scalar_pair(T=200)
    _, _, f_default, _ = gdtw.warp(x, y)
    _, _, f_explicit, _ = gdtw.warp(x, y, params={"symmetric": False})
    assert f_default == pytest.approx(f_explicit, rel=1e-12, abs=1e-12)

File: cpp/tests/test_symmetric.cpp
/* SPDX-License-Identifier: Apache-2.0
 *
 * Symmetric warping path. Mirrors python/test/test_symmetric.py.
 *
 * Asymmetric mode reads y at t_i. Symmetric mode reads y at
 * psi(t_i, j) = 2*t_i - Tau[i,j] so the residual couples both directions.
 */
#include "doctest/doctest.h"

#include "gdtw/warp.hpp"
#include "_helpers.hpp"

#include <cmath>
#include <vector>

using namespace gdtw;
using namespace test_helpers;

TEST_CASE("test_symmetric_exposes_psi_accessors") {
    auto p = scalar_pair(200);
    GDTWParams params; params.symmetric = true;
    auto r = warp(spec_from(p.x, 1), spec_from(p.y, 1), p.t, params);
    std::vector<double> grid = linspace(0.0, 1.0, 50);
    for (double g : grid) {
        CHECK(std::abs(r.g->psi(g) - (2.0 * g - r.g->phi(g))) < 1e-12);
    }
    CHECK(static_cast<int>(r.g->get_psi_values().size()) == r.g->N());
}

TEST_CASE("test_symmetric_psi_complementary") {
    auto p = scalar_pair(200);
    GDTWParams params; params.symmetric = true;
    auto r = warp(spec_from(p.x, 1), spec_from(p.y, 1), p.t, params);
    auto psi_vals = r.g->get_psi_values();
    for (int i = 0; i < r.g->N(); ++i) {
        CHECK(std::abs(psi_vals[i] + r.g->tau()[i] - 2.0 * r.g->t()[i]) < 1e-12);
    }
}

TEST_CASE("test_symmetric_default_is_asymmetric") {
    auto p = scalar_pair(200);
    auto r_def = warp(spec_from(p.x, 1), spec_from(p.y, 1), p.t);
    GDTWParams params; params.symmetric = false;
    auto r_exp = warp(spec_from(p.x, 1), spec_from(p.y, 1), p.t, params);
    CHECK(r_def.f_tau == doctest::Approx(r_exp.f_tau).epsilon(1e-12));
}
